File: test/language/functions/java/functions/structure_parent_child_value/StructureParentChildValueTest.java
package functions.structure_parent_child_value;

import static org.junit.Assert.assertEquals;
import static org.junit.Assert.assertTrue;

import java.io.IOException;
import java.util.Arrays;

import org.junit.Test;

import zserio.runtime.io.ByteArrayBitStreamReader;
import zserio.runtime.io.ByteArrayBitStreamWriter;

public class StructureParentChildValueTest
{
    @Test
    public void checkParentValue() throws IOException
    {
        final ParentValue parentValue = createParentValue();
        assertEquals(CHILD_VALUE, parentValue.funcGetValue());

        final ByteArrayBitStreamWriter writer = new ByteArrayBitStreamWriter();
        parentValue.write(writer);
        final byte[] writtenByteArray = writer.toByteArray();
        writer.close();

        final byte[] expectedByteArray = writeParentValueToByteArray();
        assertTrue(Arrays.equals(expectedByteArray, writtenByteArray));

        final ByteArrayBitStreamReader reader = new ByteArrayBitStreamReader(writtenByteArray);
        final ParentValue readParentValue = new ParentValue(reader);
        assertEquals(parentValue, readParentValue);
    }

    private byte[] writeParentValueToByteArray() throws IOException
    {
        ByteArrayBitStreamWriter writer = new ByteArrayBitStreamWriter();
        writer.writeBits(CHILD_VALUE, 32);
        writer.close();

        return writer.toByteArray();
    }

    private ParentValue createParentValue()
    {
        final ChildValue childValue = new ChildValue(CHILD_VALUE);

        return new ParentValue(childValue);
    }

    private static int CHILD_VALUE = 0xABCD;
}

File: test/language/functions/cpp98/StructureParentChildValueTest.cpp
#include <vector>

#include "gtest/gtest.h"

#include "zserio/BitStreamWriter.h"
#include "zserio/BitStreamReader.h"
#include "zserio/ObjectArray.h"

#include "functions/structure_parent_child_value/ParentValue.h"

namespace functions
{
namespace structure_parent_child_value
{

class FunctionsStructureParentChildValueTest : public ::testing::Test
{
protected:
    void writeParentValueToByteArray(zserio::BitStreamWriter& writer)
    {
        writer.writeBits(CHILD_VALUE, 32);
    }

    void createParentValue(ParentValue& parentValue)
    {
        ChildValue& childValue = parentValue.getChildValue();
        childValue.setVal(CHILD_VALUE);
        parentValue.setChildValue(childValue);
   }

    static const uint32_t CHILD_VALUE;
};

const uint32_t FunctionsStructureParentChildValueTest::CHILD_VALUE = 0xABCD;

TEST_F(FunctionsStructureParentChildValueTest, checkParentValue)
{
    ParentValue parentValue;
    createParentValue(parentValue);
    ASSERT_EQ(CHILD_VALUE, parentValue.funcGetValue());

    zserio::BitStreamWriter writtenWriter;
    parentValue.write(writtenWriter);
    size_t writtenWriterBufferByteSize;
    const uint8_t* writtenWriterBuffer = writtenWriter.getWriteBuffer(writtenWriterBufferByteSize);

    zserio::BitStreamWriter expectedWriter;
    writeParentValueToByteArray(expectedWriter);
    size_t expectedWriterBufferByteSize;
    const uint8_t* expectedWriterBuffer = expectedWriter.getWriteBuffer(expectedWriterBufferByteSize);

    std::vector<uint8_t> writtenWriterVector(writtenWriterBuffer,
                                             writtenWriterBuffer + writtenWriterBufferByteSize);
    std::vector<uint8_t> expectedWriterVector(expectedWriterBuffer,
                                              expectedWriterBuffer + expectedWriterBufferByteSize);
    ASSERT_EQ(expectedWriterVector, writtenWriterVector);

    zserio::BitStreamReader reader(writtenWriterBuffer, writtenWriterBufferByteSize);
    const ParentValue readParentValue(reader);
    ASSERT_EQ(parentValue, readParentValue);
}

} // namespace structure_parent_child_value
} // namespace functions
